### PR TITLE
Replace FunctorValueTraits for OpenMPTarget

### PR DESCRIPTION
**Pull request by masterleinad**
Part of https://github.com/kokkos/kokkos/pull/4663.

### CODE REVIEW

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 1]
```diff
@@ -623,22 +617,32 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
 #pragma omp for
         for (idx_type i = 0; i < chunk_size; ++i) {
           const idx_type idx = local_offset + i;
           value_type local_offset_value;
           if (i > 0) {
             local_offset_value = element_values(team_id, i - 1);
-            ValueJoin::join(a_functor, &local_offset_value, &offset_value);
+            // FIXME We should be able to just do
+            //   final_reducer.join(&local_offset_value, &offset_value);
+            // but we seem to memory illegaly on AMD GPUs
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:638]:
```suggestion
            // FIXME We should be able to just do
            //   final_reducer.join(&local_offset_value, &offset_value);
            // but we seem to access memory illegally on AMD GPUs
```

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 2]
```diff
@@ -867,11 +857,11 @@ class OpenMPTargetExecTeamMember {
 
   template <class ValueType, class JoinOp>
   KOKKOS_INLINE_FUNCTION ValueType team_reduce(const ValueType& value,
-                                               const JoinOp& op_in) const {
+                                               const JoinOp&) const {
 #pragma omp barrier
 
     using value_type = ValueType;
-    const JoinLambdaAdapter<value_type, JoinOp> op(op_in);
+    //    const JoinLambdaAdapter<value_type, JoinOp> op(op_in);
```
Comment (crtrott) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:864]: Why leave this commented in? Was this a bug that `op` was never used below? I don't see the team_reduce here being restricted to +?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:638]: Is it worth it maybe to ifdef based on arch define, so we do have the code we actually want here in general, but then use the workaround for AMD GPUs/

**Review by crtrott — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
Before I approve I'd like an answer to my questions, and maybe do the workaround only conditionally?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:864]: On a second glace, I'm not quite sure if `OpenMPTarget` is even implementing the right interface here. All the other backends only seem to implement
```C++
template <typename ReducerType>
KOKKOS_INLINE_FUNCTION
typename std::enable_if<is_reducer<ReducerType>::value>::type
team_reduce(ReducerType const& reducer) const noexcept;

template <typename ReducerType>
KOKKOS_INLINE_FUNCTION
typename std::enable_if<is_reducer<ReducerType>::value>::type
team_reduce(ReducerType const& reducer,
            typename ReducerType::value_type& value) const noexcept;
```
(HPX has a dummy implementation and there is an implementation for `Threads`, though).

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 3]
```diff
@@ -623,22 +617,34 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
+          final_reducer.init(&offset_value);
 
 #pragma omp for
         for (idx_type i = 0; i < chunk_size; ++i) {
           const idx_type idx = local_offset + i;
           value_type local_offset_value;
           if (i > 0) {
             local_offset_value = element_values(team_id, i - 1);
-            ValueJoin::join(a_functor, &local_offset_value, &offset_value);
+            // FIXME We seem to access memory illegaly on AMD GPUs
```
Comment (rgayatri23) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:636]: Just FYI - we are following the convention of `FIXME_OPENMPTARGET` if the issue is something related to OpenMPTarget.

**Review by rgayatri23 — APPROVED** [review]

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp [hunk 4]
```diff
@@ -70,26 +70,16 @@ template <class Reducer>
...
 
+  // Using a generic unknown Reducer for the OpenMPTarget backend is not
+  // implemented.
   KOKKOS_INLINE_FUNCTION
-  static void join(value_type&, const value_type&) {
-    printf(
-        "Using a generic unknown Reducer for the OpenMPTarget backend is not "
-        "implemented.");
-  }
+  static void join(value_type&, const value_type&) = delete;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Exec.hpp:76]: Can you clarify? The CI seems to work fine with this.

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]
Approving: but we need to sort out whats going on with the interface there

**Comment by masterleinad** [comment]
> Approving: but we need to sort out whats going on with the interface there

See #4853.

**Pull request merged**